edx_arch_experiments/datadog_diagnostics/tests/test_app.py

```diff
@@ -78,3 +80,124 @@ def test_logging(self, mock_log_error, mock_log_info):
+            # Add a bad entry to the config to test the signal lookup path
+            apps.CELERY_SIGNAL_CONFIG['fake_signal'] = lambda kwargs: {}
+            # Remove a real signal from the config to test the extractor lookup path
+            del apps.CELERY_SIGNAL_CONFIG['task_unknown']
+            apps.DatadogDiagnostics('edx_arch_experiments.datadog_diagnostics', apps).ready()
+
+        assert mock_connect.call_args_list == [
+            call(
+                celery.signals.task_success,
+                'task_success',
+                apps.CELERY_SIGNAL_CONFIG['task_success'],
+            ),
+            call(
+                celery.signals.task_rejected,
+                'task_rejected',
+                apps.CELERY_SIGNAL_CONFIG['task_rejected'],
+            ),
+        ]
+        assert mock_log_warning.call_args_list == [
+            call("Don't know how to extract info for Celery signal 'task_unknown'; ignoring."),
+            call("Could not connect receiver to unknown Celery signal 'fake_signal'"),
+        ]
+        mock_log_info.assert_called_with(
+            "Logging lifecycle info for these celery signals: ['task_success', 'task_rejected']"
+        )
+
+    @patch('edx_arch_experiments.datadog_diagnostics.apps.log.info')
+    @patch('edx_arch_experiments.datadog_diagnostics.apps.log.exception')
+    def test_handler(self, mock_log_exception, mock_log_info):
+        """Test that signal connector *behaves* as expected."""
+        # Signal A will just do a straightforward data extraction from the kwargs.
+        # pylint: disable=protected-access
+        apps._connect_celery_handler(
+            signal_example_a, 'signal_example_a',
+            lambda kwargs: {'name': kwargs['info']['name']},
+        )
+
+        # Signal B will have an extractor that goes bad on the 2nd and 3rd calls
+        b_called_times = 0
+
+        def b_extractor(kwargs):
+            nonlocal b_called_times
+            b_called_times += 1
+            if b_called_times >= 2:
+                raise Exception("oops")
+
+            return {'id': kwargs['the_id']}
+
+        # pylint: disable=protected-access
+        apps._connect_celery_handler(signal_example_b, 'signal_example_b', b_extractor)
+
+        # Send to B a few times to show that error logging only happens once
+        signal_example_b.send(sender='some_sender', the_id=42)
+        signal_example_b.send(sender='some_sender', the_id=42)
+        signal_example_b.send(sender='some_sender', the_id=42)
+        # And then send to A to show it still works
+        signal_example_a.send(
+            sender='some_sender', other='whatever', info={'name': "Alice"}, name='not this one',
+        )
+
+        mock_log_exception.assert_called_once_with(
+            "Error while extracting Celery signal info for 'signal_example_b'; "
+            "will not attempt for future calls to this signal."
```

Comment:
Probably overkill if this will all just work anyway, but consider output of the signal details if it would help debug.

Reply:
What other details would we include? I think the name is the only printable object we have our hands on at that point. (And the stack trace, which should be included since I'm using `log.exception`.)

Reply:
(Going to merge this for now but we can easily add more if needed.)

Reply:
I didn't think too deeply about this. And I forgot that this was logging the exception. All set.